File: part3.cpp
#include <iostream>
#if defined WIN32 || defined _WIN32 || defined _WIN64
#include <windows.h>
#ifndef TIME_POINT_TYPE
#define TIME_POINT_TYPE chrono::steady_clock::time_point
#endif
#else
#ifndef TIME_POINT_TYPE
#define TIME_POINT_TYPE chrono::system_clock::time_point
#endif
#endif
#include <string>
#include <sstream>
#include <fstream>
#include <vector>
#include <map>
#include <set>
#include <algorithm>
#include <chrono>
#ifndef EXIT_SUCCESS
#define EXIT_SUCCESS 0
#endif
#ifndef EXIT_FAILURE
#define EXIT_FAILURE 1
#endif
#ifndef EOF
#define EOF (-1)
#endif
#ifndef NULL
#define NULL 0
#endif
#ifndef INITIAL_VALUE
#define INITIAL_VALUE 0
#endif
#ifndef WINDOWS_MAX_PATH
#define WINDOWS_MAX_PATH 32762
#endif
using namespace std;
char path[WINDOWS_MAX_PATH + 1] = { 0 };


struct Point
{
	const size_t index = INITIAL_VALUE;
	const long double x = INITIAL_VALUE, y = INITIAL_VALUE;
};


class KeywordLocationSearch
{
private:
	vector<string> lines{};
	map<string, set<size_t>> dict{};
	vector<vector<vector<Point>>> cells{};
	long int gridSize = 0; // 0 for non-initialization
	long double xLB = INITIAL_VALUE, xUB = INITIAL_VALUE, yLB = INITIAL_VALUE, yUB = INITIAL_VALUE, xSize = INITIAL_VALUE, ySize = INITIAL_VALUE;
	
	bool getTxt(const string filePath)
	{
		try
		{
			ifstream fp = ifstream(filePath);
			if (fp.is_open())
			{
				string line{};
				while (!fp.eof())
				{
					getline(fp, line);
					this->lines.push_back(line);
				}
				if (!this->lines.empty() && this->lines[this->lines.size() - 1].empty()) // remove the empty line at the end of the file
					this->lines.pop_back();
				return true;
			}
			else
				return false;
		}
		catch (...)
		{
			return false;
		}
	}
	set<size_t> kwSearchIF(int argc, char* argv[])
	{
		if (argc >= 6)
		{
			set<size_t> s = this->dict.find(argv[5]) != this->dict.end() ? this->dict[argv[5]] : set<size_t>{};
			for (int i = 6; i < argc && !s.empty(); ++i)
				if (this->dict.find(argv[i]) != this->dict.end())
				{
					set<size_t> tmp{};
					set_intersection(s.begin(), s.end(), this->dict[argv[i]].begin(), this->dict[argv[i]].end(), inserter(tmp, tmp.end()));
					s = tmp;
				}
				else
				{
					s.clear();
					break;
				}
			return s;
		}
		else if (5 == argc)
		{
			set<size_t> s{};
			const size_t length = this->lines.size();
			for (size_t idx = 0; idx < length; ++idx)
				s.insert(idx);
			return s;
		}
		else
			return set<size_t>{};
	}
	vector<size_t> spaSearchGrid(long double xLow, long double xHigh, long double yLow, long double yHigh)
	{
		vector<size_t> results{};
		if (this->gridSize >= 1)
		{
			const long int xLowIdx = this->xLB < xLow && xLow < this->xUB ? (long int)((xLow - this->xLB) / this->xSize) : (xLow <= this->xLB ? -1 : this->gridSize);
			const long int xHighIdx = this->xLB < xHigh && xHigh < this->xUB ? (long int)((xHigh - this->xLB) / this->xSize) : (xHigh <= this->xLB ? -1 : this->gridSize);
			const long int yLowIdx = this->yLB < yLow && yLow < this->yUB ? (long int)((yLow - this->yLB) / this->ySize) : (yLow <= this->yLB ? -1 : this->gridSize);
			const long int yHighIdx = this->yLB < yHigh && yHigh < this->yUB ? (long int)((yHigh - this->yLB) / this->ySize) : (yHigh <= this->yLB ? -1 : this->gridSize);
			if (0 <= yHighIdx && yHighIdx < this->gridSize) // the upper side (from left to right and exclude the rightmost cell)
				for (long int i = max((long int)0, xLowIdx); i < min(xHighIdx, this->gridSize - 1); ++i)
					for (const Point& point : this->cells[i][yHighIdx])
						if (xLow <= point.x && point.x <= xHigh && yLow <= point.y && point.y <= yHigh)
							results.push_back(point.index);
			if (0 <= xHighIdx && xHighIdx < this->gridSize) // the right side (from up to bottom and exclude the bottom cell)
				for (long int j = min(yHighIdx, this->gridSize - 1); j > max((long int)0, yLowIdx); --j)
					for (const Point& point : this->cells[xHighIdx][j])
						if (xLow <= point.x && point.x <= xHigh && yLow <= point.y && point.y <= yHigh)
							results.push_back(point.index);
			if (0 <= yLowIdx && yLowIdx < this->gridSize) // the lower side (from right to left and exclude the leftmost cell)
				for (long int i = min(xHighIdx, this->gridSize - 1); i > max((long int)0, xLowIdx); --i)
					for (const Point& point : this->cells[i][yLowIdx])
						if (xLow <= point.x && point.x <= xHigh && yLow <= point.y && point.y <= yHigh)
							results.push_back(point.index);
			if (0 <= xLowIdx && xLowIdx < this->gridSize) // the left side (from bottom to up and exclude the top cell)
				for (long int j = max((long int)0, yLowIdx); j < min(yHighIdx, this->gridSize - 1); ++j)
					for (const Point& point : this->cells[xLowIdx][j])
						if (xLow <= point.x && point.x <= xHigh && yLow <= point.y && point.y <= yHigh)
							results.push_back(point.index);
			for (long int i = xLowIdx + 1; i < xHighIdx; ++i) // all the points in the cells within but not on the four sides must be within the given range
				for (long int j = yLowIdx + 1; j < yHighIdx; ++j)
					for (const Point& point : this->cells[i][j])
						results.push_back(point.index);
			sort(results.begin(), results.end());
		}
		return results;
	}
	
public:
	static const size_t INVALID = (size_t)(-1);
	
	KeywordLocationSearch() {}
	bool build(const string filePath, long int gSize)
	{
		/* Initialization */
		if (gSize < 1 || !this->getTxt(filePath))
			return false;
		this->dict.clear();
		this->cells.clear(); // cells are also cleared here since the private variable ``dict`` will be modified during the construction
		
		/* Construction */
		vector<Point> points{};
		const size_t lineCount = this->lines.size();
		for (size_t i = 0; i < lineCount; ++i)
		{
			const size_t locationPos = this->lines[i].find("\tlocation: "), tagsPos = this->lines[i].find("\ttags: ");
			if (string::npos != locationPos && string::npos != tagsPos)
			{
				/* Tags */
				stringstream tags(lines[i].substr(tagsPos + 7));
				while (!tags.eof())
				{
					string tag{};
					getline(tags, tag, ',');
					if (this->dict.find(tag) == this->dict.end())
						this->dict[tag] = set<size_t>{ i };
					else
						this->dict[tag].insert(i);
				}
				
				/* Location Grid */
				const string subString = this->lines[i].substr(locationPos + 11, tagsPos - locationPos - 11);
				const size_t lIdx = subString.find(","), rIdx = subString.rfind(",");
				if (string::npos != lIdx && lIdx == rIdx) // to guarantee that there is only a ','
				{
					try
					{
						points.push_back(Point{ i, stold(subString.substr(0, lIdx)), stold(subString.substr(lIdx + 1)) });
					}
					catch (...) {}
				}
			}
		}
		if (points.empty())
		{
			this->gridSize = 0; // 0 for non-initialization
			return false;
		}
		
		/* Bounds */
		this->xLB = ((vector<Point>::iterator)min_element(points.begin(), points.end(), [](const Point& a, const Point& b) { return a.x < b.x; }))->x;
		this->xUB = ((vector<Point>::iterator)max_element(points.begin(), points.end(), [](const Point& a, const Point& b) { return a.x < b.x; }))->x;
		this->yLB = ((vector<Point>::iterator)min_element(points.begin(), points.end(), [](const Point& a, const Point& b) { return a.y < b.y; }))->y;
		this->yUB = ((vector<Point>::iterator)max_element(points.begin(), points.end(), [](const Point& a, const Point& b) { return a.y < b.y; }))->y;
		const long double xDelta = xUB - xLB, yDelta = yUB - yLB;
		
		/* Cells */
		for (long int i = 0; i < gSize; ++i)
		{
			this->cells.push_back(vector<vector<Point>>{});
			for (long int j = 0; j < gSize; ++j)
				this->cells[i].push_back(vector<Point>{});
		}
		this->xSize = xDelta / gSize;
		this->ySize = yDelta / gSize;
		for (const Point& point : points)
			this->cells[min(gSize - 1, (long int)((point.x - xLB) / xSize))][min(gSize - 1, (long int)((point.y - yLB) / ySize))].push_back(point);
		this->gridSize = gSize;
		return true;
	}
	bool build(const string filePath) { return this->build(filePath, 50); }
	size_t kwSpaSearchRaw(long double xLow, long double xHigh, long double yLow, long double yHigh, int argc, char* argv[])
	{
		if (this->gridSize >= 1 && argc >= 5)
		{
			vector<size_t> results{};
			const size_t lineCount = this->lines.size();
			const TIME_POINT_TYPE startTime = chrono::high_resolution_clock::now();
			for (size_t idx = 0; idx < lineCount; ++idx)
			{
				const size_t locationPos = this->lines[idx].find("\tlocation: "), tagsPos = this->lines[idx].find("\ttags: ");
				if (string::npos != locationPos && string::npos != tagsPos)
				{
					/* Keyword */
					vector<string> tags{};
					stringstream ss(this->lines[idx].substr(tagsPos + 7));
					while (!ss.eof())
					{
						string tag{};
						getline(ss, tag, ',');
						tags.push_back(tag);
					}
					bool flag = true;
					for (int i = 5; i < argc; ++i)
						if (find(tags.begin(), tags.end(), argv[i]) == tags.end())
						{
							flag = false;
							break;
						}
					if (flag)
					{
						/* Location */
						const string subString = this->lines[idx].substr(locationPos + 11, tagsPos - locationPos - 11);
						const size_t lIdx = subString.find(","), rIdx = subString.rfind(",");
						if (string::npos != lIdx && lIdx == rIdx) // to guarantee that there is only a ','
						{
							long double x = INITIAL_VALUE, y = INITIAL_VALUE;
							try
							{
								x = stold(subString.substr(0, lIdx));
								y = stold(subString.substr(lIdx + 1));
							}
							catch (...)
							{
								continue;
							}
							if (xLow <= x && x <= xHigh && yLow <= y && y <= yHigh)
								results.push_back(idx);
						}
					}
				}
			}
			const long double timeDelta = (long double)(chrono::high_resolution_clock::now() - startTime).count() / 1000000000;
			cout << "kwSpaSearchRaw: " << results.size() << " result(s), cost = " << fixed << timeDelta << " second(s)" << endl;
			for (const size_t& idx : results)
				cout << this->lines[idx] << endl;
			cout << endl;
			return results.size();
		}
		else
			return INVALID;
	}
	size_t kwSpaSearchIF(long double xLow, long double xHigh, long double yLow, long double yHigh, int argc, char* argv[])
	{
		if (this->gridSize >= 1 && argc >= 5)
		{
			vector<size_t> results{};
			const TIME_POINT_TYPE startTime = chrono::high_resolution_clock::now();
			set<size_t> s = this->kwSearchIF(argc, argv);
			for (const size_t& idx : s)
			{
				const size_t locationPos = this->lines[idx].find("\tlocation: "), tagsPos = this->lines[idx].find("\ttags: ");
				if (string::npos != locationPos && string::npos != tagsPos)
				{
					const string subString = this->lines[idx].substr(locationPos + 11, tagsPos - locationPos - 11);
					const size_t lIdx = subString.find(","), rIdx = subString.rfind(",");
					if (string::npos != lIdx && lIdx == rIdx) // to guarantee that there is only a ','
					{
						long double x = INITIAL_VALUE, y = INITIAL_VALUE;
						try
						{
							x = stold(subString.substr(0, lIdx));
							y = stold(subString.substr(lIdx + 1));
						}
						catch (...)
						{
							continue;
						}
						if (xLow <= x && x <= xHigh && yLow <= y && y <= yHigh)
							results.push_back(idx);
					}
				}
			}
			const long double timeDelta = (long double)(chrono::high_resolution_clock::now() - startTime).count() / 1000000000;
			cout << "kwSpaSearchIF: " << results.size() << " result(s), cost = " << fixed << timeDelta << " second(s)" << endl;
			for (const size_t& idx : results)
				cout << this->lines[idx] << endl;
			cout << endl;
			return results.size();
		}
		else
			return INVALID;
	}
	size_t kwSpaSearchGrid(long double xLow, long double xHigh, long double yLow, long double yHigh, int argc, char* argv[])
	{
		if (this->gridSize >= 1 && argc >= 5)
		{
			const TIME_POINT_TYPE startTime = chrono::high_resolution_clock::now();
			vector<size_t> vec = this->spaSearchGrid(xLow, xHigh, yLow, yHigh), results{};
			for (const size_t& idx : vec)
			{
				const size_t pos = this->lines[idx].find("\ttags: ");
				if (string::npos != pos)
				{
					vector<string> tags{};
					stringstream ss(this->lines[idx].substr(pos + 7));
					while (!ss.eof())
					{
						string tag{};
						getline(ss, tag, ',');
						tags.push_back(tag);
					}
					bool flag = true;
					for (int i = 5; i < argc; ++i)
						if (find(tags.begin(), tags.end(), argv[i]) == tags.end())
						{
							flag = false;
							break;
						}
					if (flag)
						results.push_back(idx);
				}
			}
			const long double timeDelta = (long double)(chrono::high_resolution_clock::now() - startTime).count() / 1000000000;
			cout << "kwSpaSearchGrid: " << results.size() << " result(s), cost = " << fixed << timeDelta << " second(s)" << endl;
			for (const size_t& idx : results)
				cout << this->lines[idx] << endl;
			cout << endl;
			return results.size();
		}
		else
			return INVALID;
	}
};


#if defined WIN32 || defined _WIN32 || defined _WIN64
static bool cdCurrentDirectory()
{
	GetModuleFileNameA(NULL, path, WINDOWS_MAX_PATH);
	const string::size_type pos = string(path).find_last_of("\\/");
	if (string::npos != pos)
	{
		const string dir = string(path).substr(0, pos + 1); // to remain the path separator to avoid paths like "C:\\"
		SetCurrentDirectoryA(dir.c_str());
		GetCurrentDirectoryA(MAX_PATH, path);
#if _DEBUG
		cout << "The working directory has been changed to \"" << path << "\". " << endl;
#endif
		return true;
	}
	else
		return false;
}
#endif

static int pressAnyKeyToContinue()
{
	rewind(stdin);
	fflush(stdin);
	cout << "The program is about to exit. Please press any key to continue. " << endl;
	return getchar();
}



int main(int argc, char* argv[])
{
	long double xLow = INITIAL_VALUE, xHigh = INITIAL_VALUE, yLow = INITIAL_VALUE, yHigh = INITIAL_VALUE;
	if (argc >= 5) // 51 51.50 -0.5 0 british bar
	{
		try
		{
			xLow = stold(argv[1]);
			xHigh = stold(argv[2]);
			yLow = stold(argv[3]);
			yHigh = stold(argv[4]);
		}
		catch (...)
		{
			cout << "Please provide the four boundary values in the form of ``xLow xHigh yLow yHigh`` correctly before the keywords. " << endl;
			pressAnyKeyToContinue();
			return EOF;
		}
	}
	else
	{
		cout << "Please provide the locations and the keywords for searching via the command line. " << endl;
		pressAnyKeyToContinue();
		return EOF;
	}
#if defined WIN32 || defined _WIN32 || defined _WIN64
	cdCurrentDirectory();
	const UINT currentConsoleOutputCP = GetConsoleOutputCP(); // backup the active code page
	if (!SetConsoleOutputCP(65001))
		cout << "Warning: As the file is encoded by UTF-8 while the program fails to change the active code page of the current console to 65001, the display may be incorrect. " << endl;
#endif
	KeywordLocationSearch keywordLocationSearch{};
	if (!(keywordLocationSearch.build("Restaurants_London_England.tsv") || keywordLocationSearch.build("../../Restaurants_London_England.tsv") || keywordLocationSearch.build("../../../Restaurants_London_England.tsv")))
	{
		cout << "Cannot read the file \"Restaurants_London_England.tsv\". " << endl;
		cout << "Please also make sure that the grid size passed is a positive integer. " << endl;
		pressAnyKeyToContinue();
		return EOF;
	}
	const size_t flagRaw = keywordLocationSearch.kwSpaSearchRaw(xLow, xHigh, yLow, yHigh, argc, argv), flagIF = keywordLocationSearch.kwSpaSearchIF(xLow, xHigh, yLow, yHigh, argc, argv), flagGrid = keywordLocationSearch.kwSpaSearchGrid(xLow, xHigh, yLow, yHigh, argc, argv);
#if defined WIN32 || defined _WIN32 || defined _WIN64
	SetConsoleOutputCP(currentConsoleOutputCP); // restore the active code page
#endif
	pressAnyKeyToContinue();
	return KeywordLocationSearch::INVALID != flagRaw && KeywordLocationSearch::INVALID != flagIF && KeywordLocationSearch::INVALID != flagGrid && flagRaw > 0 && flagIF >= 0 && flagGrid > 0 && flagRaw == flagIF && flagRaw == flagGrid? EXIT_SUCCESS : EXIT_FAILURE;
}

File: part3.py
import os
from sys import argv, exit
from time import time, sleep
try:
	os.chdir(os.path.abspath(os.path.dirname(__file__)))
except:
	pass
EXIT_SUCCESS = 0
EXIT_FAILURE = 1
EOF = (-1)


class KeywordLocationSearch:
	def __init__(self:object) -> object:
		self.__lines = []
		self.__dict = {}
		self.__cells = []
		self.__gridSize = 0 # flag
		self.__xLB = None
		self.__xUB = None
		self.__yLB = None
		self.__yUB = None
		self.__xSize = None
		self.__ySize = None
	def __getTxt(self:object, filePath:str, index:int = 0) -> str: # get .txt content
		coding = ("utf-8", "gbk", "utf-16") # codings
		if 0 <= index < len(coding): # in the range
			try:
				with open(filePath, "r", encoding = coding[index]) as f:
					content = f.read()
				return content[1:] if content.startswith("\ufeff") else content # if utf-8 with BOM, remove BOM
			except (UnicodeError, UnicodeDecodeError):
				return getTxt(filePath, index + 1) # recursion
			except:
				return None
		else:
			return None # out of range
	def build(self:object, filePath:str, gridSize:int = 50) -> bool:
		# Initialization #
		if isinstance(filePath, str) and isinstance(gridSize, int) and gridSize >= 1:
			content = self.__getTxt(filePath)
			if content is None:
				print("Cannot read the file \"{0}\". ".format(filePath))
				return False # do not set the gridSize to 0 (set the flag to False) since no modifications are made in the lines and the cells
			else:
				lines = content.splitlines()
				self.__lines = lines
				del content
		else:
			print("Please pass the file path as a ``str`` object and the grid size as a positive integer. ")
			return False
		
		# Construction #
		d = self.__dict
		d.clear()
		tuples = []
		for i, line in enumerate(lines):
			if "\tlocation: " in line and "\ttags: " in line:
				tags = line[line.index("\ttags: ") + 7:].split(",")
				for tag in tags:
					d.setdefault(tag, set()).add(i)
				values = line[line.index("\tlocation: ") + 11:line.index("\ttags: ")].split(",")
				if 2 == len(values):
					try:
						tuples.append((i, float(values[0]), float(values[1])))
					except:
						pass
		if not tuples:
			self.__gridSize = 0 # set the flag to False to indicate that the object is not ready
			return False
		
		# Bounds #
		xLB, xUB, yLB, yUB = min(tuples, key = lambda t:t[1])[1], max(tuples, key = lambda t:t[1])[1], min(tuples, key = lambda t:t[2])[2], max(tuples, key = lambda t:t[2])[2]
		xDelta, yDelta = xUB - xLB, yUB - yLB
		
		# Cells #
		self.__cells = [[[] for j in range(gridSize)] for i in range(gridSize)]
		cells = self.__cells # to speed up
		xSize, ySize = xDelta / gridSize, yDelta / gridSize
		for t in tuples:
			cells[min(gridSize - 1, int((t[1] - xLB) / xSize))][min(gridSize - 1, int((t[2] - yLB) / ySize))].append(t)
		
		# Variables #
		self.__gridSize, self.__xLB, self.__xUB, self.__yLB, self.__yUB, self.__xSize, self.__ySize = gridSize, xLB, xUB, yLB, yUB, xSize, ySize
		return True
	def kwSpaSearchRaw(self:object, xLow:float, xHigh:float, yLow:float, yHigh:float, keywords:list) -> int:
		if isinstance(keywords, (tuple, list, set)) and isinstance(xLow, float) and isinstance(xHigh, float) and isinstance(yLow, float) and isinstance(yHigh, float) and self.__gridSize >= 1:
			lines, results = self.__lines, []
			startTime = time()
			for i, line in enumerate(lines):
				if "\tlocation: " in line and "\ttags: " in line:
					tags = line[line.index("\ttags: ") + 7:].split(",")
					values = line[line.index("\tlocation: ") + 11:line.index("\ttags: ")].split(",")
					if 2 == len(values):
						try:
							x, y = float(values[0]), float(values[1])
						except:
							continue
						for keyword in keywords:
							if keyword not in tags:
								break
						else: # the loop ends naturally
							if xLow <= x <= xHigh and yLow <= y <= yHigh:
								results.append(i)
			endTime = time()
			print("kwSpaSearchRaw: {0} result(s), cost = {1:.9f} second(s)".format(len(results), endTime - startTime))
			for idx in results:
				print(lines[idx])
			print()
			return len(results)
		else:
			return -1
	def __kwSearchIF(self:object, keywords:list) -> list:
		if isinstance(keywords, (tuple, list, set)):
			lines, d = self.__lines, self.__dict
			if keywords:
				idx, length, s = 1, len(keywords), (d[keywords[0]] if keywords[0] in d else set())
				while idx < length and s: # to speed up
					if keywords[idx] in d: # avoid bugs
						s = s.intersection(d[keywords[idx]])
						idx += 1
					else:
						s.clear()
						break
				return sorted(s)
			else:
				return list(range(len(lines)))
		else:
			return []
	def kwSpaSearchIF(self:object, xLow:float, xHigh:float, yLow:float, yHigh:float, keywords:list) -> int:
		if isinstance(keywords, (tuple, list, set)) and isinstance(xLow, float) and isinstance(xHigh, float) and isinstance(yLow, float) and isinstance(yHigh, float) and self.__gridSize >= 1:
			lines = self.__lines
			startTime = time()
			vec, results = self.__kwSearchIF(keywords), []
			for idx in vec:
				line = lines[idx]
				if "\tlocation: " in line and "\ttags: " in line:
					values = line[line.index("\tlocation: ") + 11:line.index("\ttags: ")].split(",")
					if 2 == len(values):
						try:
							x, y = float(values[0]), float(values[1])
						except:
							continue
						if xLow <= x <= xHigh and yLow <= y <= yHigh:
							results.append(idx)
			endTime = time()
			print("kwSpaSearchIF: {0} result(s), cost = {1:.9f} second(s)".format(len(results), endTime - startTime))
			for idx in results:
				print(lines[idx])
			print()
			return len(results)
		else:
			return -1
	def __spaSearchGrid(self:object, xLow:float, xHigh:float, yLow:float, yHigh:float) -> list:
		if isinstance(xLow, float) and isinstance(xHigh, float) and isinstance(yLow, float) and isinstance(yHigh, float) and self.__gridSize >= 1:
			lines, cells, results, gridSize, xLB, xUB, yLB, yUB, xSize, ySize = self.__lines, self.__cells, [], self.__gridSize, self.__xLB, self.__xUB, self.__yLB, self.__yUB, self.__xSize, self.__ySize
			xLowIdx = int((xLow - xLB) / xSize) if xLB < xLow < xUB else (-1 if xLow <= xLB else gridSize)
			xHighIdx = int((xHigh - xLB) / xSize) if xLB < xHigh < xUB else (-1 if xHigh <= xLB else gridSize)
			yLowIdx = int((yLow - yLB) / ySize) if yLB < yLow < yUB else (-1 if yLow <= yLB else gridSize)
			yHighIdx = int((yHigh - yLB) / ySize) if yLB < yHigh < yUB else (-1 if yHigh <= yLB else gridSize)
			if 0 <= yHighIdx < gridSize: # the upper side (from left to right and exclude the rightmost cell)
				for i in range(max(0, xLowIdx), min(xHighIdx, gridSize - 1)):
					for t in cells[i][yHighIdx]:
						if xLow <= t[1] <= xHigh and yLow <= t[2] <= yHigh:
							results.append(t[0])
			if 0 <= xHighIdx < gridSize: # the right side (from up to bottom and exclude the bottom cell)
				for j in range(min(yHighIdx, gridSize - 1), max(0, yLowIdx), -1):
					for t in cells[xHighIdx][j]:
						if xLow <= t[1] <= xHigh and yLow <= t[2] <= yHigh:
							results.append(t[0])
			if 0 <= yLowIdx < gridSize: # the lower side (from right to left and exclude the leftmost cell)
				for i in range(min(xHighIdx, gridSize - 1), max(0, xLowIdx), -1):
					for t in cells[i][yLowIdx]:
						if xLow <= t[1] <= xHigh and yLow <= t[2] <= yHigh:
							results.append(t[0])
			if 0 <= xLowIdx < gridSize: # the left side (from bottom to up and exclude the top cell)
				for j in range(max(0, yLowIdx), min(yHighIdx, gridSize - 1)):
					for t in cells[xLowIdx][j]:
						if xLow <= t[1] <= xHigh and yLow <= t[2] <= yHigh:
							results.append(t[0])
			for i in range(xLowIdx + 1, xHighIdx): # all the points in the cells within but not on the four sides must be within the given range
				for j in range(yLowIdx + 1, yHighIdx):
					for t in cells[i][j]:
						results.append(t[0])
			return sorted(results)
		else:
			return []
	def kwSpaSearchGrid(self:object, xLow:float, xHigh:float, yLow:float, yHigh:float, keywords:list) -> int:
		if isinstance(keywords, (tuple, list, set)) and isinstance(xLow, float) and isinstance(xHigh, float) and isinstance(yLow, float) and isinstance(yHigh, float) and self.__gridSize >= 1:
			lines = self.__lines
			startTime = time()
			vec, results = self.__spaSearchGrid(xLow, xHigh, yLow, yHigh), []
			for idx in vec:
				line = lines[idx]
				if "\ttags: " in line:
					tags = line[line.index("\ttags: ") + 7:].split(",")
					for keyword in keywords:
						if keyword not in tags:
							break
					else: # the loop ends naturally
						results.append(idx)
			endTime = time()
			print("kwSpaSearchGrid: {0} result(s), cost = {1:.9f} second(s)".format(len(results), endTime - startTime))
			for idx in results:
				print(lines[idx])
			print()
			return len(results)
		else:
			return -1


def preExit(countdownTime:int = 5) -> None:
		try:
			cntTime = int(countdownTime)
			length = len(str(cntTime))
		except:
			print("Program ended. ")
			return
		print()
		while cntTime > 0:
			print("\rProgram ended, exiting in {{0:>{0}}} second(s). ".format(length).format(cntTime), end = "")
			try:
				sleep(1)
			except:
				print("\rProgram ended, exiting in {{0:>{0}}} second(s). ".format(length).format(0))
				return
			cntTime -= 1
		print("\rProgram ended, exiting in {{0:>{0}}} second(s). ".format(length).format(cntTime))

def main() -> int:
	if len(argv) >= 5: # 51 51.50 -0.5 0 british bar
		try:
			xLow, xHigh, yLow, yHigh = float(argv[1]), float(argv[2]), float(argv[3]), float(argv[4])
		except:
			print("Please provide the four boundary values in the form of ``xLow xHigh yLow yHigh`` correctly before the keywords. ")
			preExit()
			return EOF
	else:
		print("Please provide the locations and the keywords for searching via the command line. ")
		preExit()
		return EOF
	filePath = "Restaurants_London_England.tsv"
	keywordLocationSearch = KeywordLocationSearch()
	if not keywordLocationSearch.build(filePath):
		preExit()
		return EOF
	flagRaw = keywordLocationSearch.kwSpaSearchRaw(xLow, xHigh, yLow, yHigh, argv[5:])
	flagIF = keywordLocationSearch.kwSpaSearchIF(xLow, xHigh, yLow, yHigh, argv[5:])
	flagGrid = keywordLocationSearch.kwSpaSearchGrid(xLow, xHigh, yLow, yHigh, argv[5:])
	preExit()
	return EXIT_SUCCESS if flagRaw >= 1 and flagIF >= 1 and flagGrid >= 1 and flagRaw == flagIF == flagGrid else EXIT_FAILURE



if __name__ == "__main__":
	exit(main())